FaunaDB.Client/Encoding/Encoder.cs

```diff
@@ -0,0 +1,244 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using FaunaDB.Query;
+using FaunaDB.Types;
+using FaunaDB.Attributes;
+using System.Linq.Expressions;
+
+namespace FaunaDB.Encoding
+{
+    public static class Encoder
+    {
+        public static Expr Encode(object obj) =>
+            new EncoderImpl().Encode(obj);
+    }
+
+    class EncoderImpl
+    {
+        static readonly Dictionary<Type, Func<object, Expr>> converters = new Dictionary<Type, Func<object, Expr>>();
+
+        Stack<object> stack = new Stack<object>();
+
+        public Expr Encode(object obj)
+        {
+            if (obj == null)
+                return NullV.Instance;
+
+            if (typeof(Expr).IsInstanceOfType(obj))
+                return (Expr)obj;
+
+            if (stack.Contains(obj))
```

Comment:
Under the hood it will use an `EqualityComparer` that will check if the type implements `IEqualiable` if not it will use `Equals()` and `GetHashCode()`, thinking better now it might be an issue because if we have have an object that implements `Equals()` and `GetHashCode()` and it happens to be true for different instances it will throw the exception. I'll use a overloaded version of `Contains` that receive an custom comparer to check only for references